feat: expand bounding boxes to squares for better object detection

Karanveer266 opened the pull request:
## Square Bounding Box Implementation

This PR improves the object detection pipeline by implementing square bounding boxes for object crops.

### Changes

- Added `expand_bbox_to_square` method to transform rectangular boxes to squares
- Modified `write_crops` to use square crops while preserving object centers
- Implemented boundary handling for boxes near image edges

### Benefits

- Maintains original object proportions without artificial padding
- Preserves natural context around objects for better classification
- Reduces misidentification by providing consistent square inputs
- Handles edge cases gracefully when objects are near image boundaries

The implementation expands bounding boxes by extending the shorter dimension while keeping the original center point, ensuring we capture the complete object with its natural surroundings.

henrykironde reviewed (changes requested):
@Karanveer266 We should have some tests for this new function

ethanwhite commented:
@Karanveer266 - the test failures were independent of this PR and should all be fixed now. If you can rebase the PR against the main branch the tests should pass

Karanveer266 commented:
Hi @ethanwhite, thanks for the heads-up! I've just rebased this PR onto the latest main branch.

ethanwhite commented:
Also, to clarify @henrykironde's request above, @Karanveer266 can you please add at least one unit test for your new function as well. Once we have a test and the tests are passing we'll review the PR. Please add a comment after you add the test to make sure we see it. Thanks!

Karanveer266 commented:
Sure @ethanwhite @henrykironde , I will add the test soon.

ethanwhite commented:
Thanks @Karanveer266. Can you also remove the code style and documentation changes from this PR. Those should be submitted separately as it makes reviewing the PR a lot easier when only the key changes are included.

Karanveer266 commented:
@ethanwhite @henrykironde I have added two case tests for the expand_bbox_to_square function:

- **Test 1:** Bounding box at the top-left edge (0,0) to verify proper square expansion without exceeding image boundaries.

- **Test 2:** Bounding box larger than image dimensions to confirm the result is correctly clamped to the image size.

I have also removed the code style and documentation changes from this PR. Please go through the changes and let me know the changes required here.

Thanks for you guidance!

ethanwhite commented on src/deepforest/model.py:
```diff
@@ -209,12 +253,20 @@ def write_crops(self, root_dir, images, boxes, labels, savedir):
+                
+                # Expand the bounding box to a square
+                xmin, ymin, xmax, ymax = box
+                square_box = self.expand_bbox_to_square([xmin, ymin, xmax, ymax], image_width, image_height)
```
According to the docs `boxes` it looks like boxes is a nested list in which case we wouldn't need to split it into individual objects and then reassemble it. So these two lines could be come:

```python
square_box = self.expand_bbox_to_square(box, image_width, image_height)
```

But I haven't explicitly checked.

ethanwhite commented on tests/test_model.py:
```diff
@@ -168,3 +168,25 @@ def test_crop_model_load_checkpoint_with_explicit_num_classes(tmpdir, crop_model
+    expected = [0.0, 0.0, 100.0, 100.0]
+    result = crop_model.expand_bbox_to_square(bbox, image_width, image_height)
+    assert result == expected
+
```
This is great. Can we add one more basic test where there are no edges that influence the expansion

ethanwhite reviewed (changes requested):
This looks really good @Karanveer266! Thanks! I've asked for one more test case to be added and made one suggestion for code simplification for you to check.

Once you've made those changes I'd like to ask for two things to help us maintain a clean readable git history:
1. Rebase the entire PR into a single commit
2. Expand the commit message to provide additional information. The primary commit line is already really good, but let's remove the `feat` tag since we don't use it so and then add in the idea that this is for the crop model. To keep it to the recommend 50 characters I'd probably use: "Expand bounding boxes to squares in crop model". Then take some of the beautiful detailed information from the PR comment and include that in the body of the commit message.

Here's a good resource for thinking about how to write great commit histories: 
https://cbea.ms/git-commit/

Once that's all done this should be ready to merge!

ethanwhite commented:
@Karanveer266 - looks like there's also a style test failing. You can run the following command from the command line to clean it up:

```sh
yapf -i --recursive src/deepforest/ --style=.style.yapf
```

See docs here: https://deepforest.readthedocs.io/en/v1.5.0/development/contributing.html#using-yapf

ethanwhite commented:
Fixed in https://huggingface.co/weecology/deepforest-livestock/commit/5d13e1d8f4b4b786e9a2251f0011a85c395a985a

Karanveer266 commented:
Thanks for pointing that out @ethanwhite , I added the basic test case and added 
`square_box = self.expand_bbox_to_square(box, image_width, image_height)` in `model.py`. I also fixed the style using yapf command. I have also rebased the entire PR into a single commit. Let me know if there is anything else I need to change in this PR.
Thanks for you guidance!

ethanwhite reviewed (approved):
Thanks again for all of your work on this @Karanveer266. This is ready to merge!

Karanveer266 commented:
Glad I could help :)
